apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -64,6 +74,50 @@ type CompositionSpec struct {
 	WriteConnectionSecretsToNamespace *string `json:"writeConnectionSecretsToNamespace,omitempty"`
 }
 
+// InlinePatchSets dereferences PatchSets and includes their patches inline. The
+// updated CompositionSpec should not be persisted to the API server.
+func (cs *CompositionSpec) InlinePatchSets() error {
+	pn := make(map[string][]Patch)
+	for _, s := range cs.PatchSets {
+		for _, p := range s.Patches {
+			if p.Type == PatchTypePatchSet {
+				return errors.New(errPatchSetType)
+			}
+		}
+		pn[s.Name] = s.Patches
+	}
+
+	for i, r := range cs.Resources {
+		po := []Patch{}
```

Comment:
@negz I've changed this slightly from the sketch, as `PatchSet` dereferenced patches would always be applied in-order _after_ all of the normal patches applied on the resource. 

This way maintains order between the standard patches and the ordering of the patches inside the referenced `PatchSets`, so e.g. you could have a `PatchSet` which copies all of `metadata.annotations` from the `Composition` and then have a later `Patch` definition which overwrites a single annotation.

Reply:
Good call - it makes sense to me to keep the ordering as close to original as possible.